Objects/dictobject.c

```diff
@@ -1767,6 +1778,10 @@ _PyDict_Pop_KnownHash(PyObject *dict, PyObject *key, Py_hash_t hash, PyObject *d
     Py_DECREF(old_key);
 
     assert(_PyDict_CheckConsistency(mp));
+    if (after_removed(mp) < 0) {
+        Py_DECREF(old_value);
```

Comment:
I'm not sure that it's ok to drop the popped value here :-(

If shrinking the dict fails, I suggest to ignore the error and return the old value anyway. The next del/pop will try again to shrink the dict.